### DISCUSSION TITLE
Cloudinary (Drive) Storage Support

### DISCUSSION THREAD

**Discussion by don41382** [directus/directus#10254] [Feature Requests]
I am currently working for a client, which would like to use directus. They use Cloudinary as their storage layer.

Sadly, directus doesn't support it. I would like to add a drive implementation. I am not a nodeJS expert, but it shouldn't be that hard.

Wdyt? Any recommendation for the start?

**Comment by rijkvanzanten** [comment]
Heya! 

Love the idea, and like you said, the basic functionality shouldn't be too tricky. For starters, I would probably copy paste the drive-s3 version, and start altering every method to use Cloudinary's APIs / SDKs instead of S3:

https://github.com/directus/directus/blob/e7bfafede95254b7b1cf86e443ed75ee9159d74b/packages/drive-s3/src/AmazonWebServicesS3Storage.ts#L202-L215

However, there are a couple trickier questions that will arise around file organization specifically. Seeing that Cloudinary itself offers a DAM, we'll have to figure out how to do the two-way binding, if that's at all in scope. If you change a file's path in Cloudinary, we wouldn't want Directus to break and vise-versa. Secondly, Directus' file-management on the storage adapter might not be the ideal way to store files within Cloudinary itself. For example, we use the UUID of the file as the file name, to prevent naming conflicts in the folder. Similarly, we don't currently sync the Directus Folders setup with the storage adapters.

I'm down to implement Cloudinary now, and treat it as any other storage adapter (meaning that we treat it as "just" another folder) and look into ways to optimize the integration later. 

Curious to hear your thoughts!

**Reply by rijkvanzanten** [reply]
Oh! And another thing to keep in mind that Directus has it's own asset-generation endpoints, that will create and save the generated thumbnail on the storage adapter, this might cause some noise if you use another interface to explore these files (like Cloudinary). We might want to add an option to specify a folder path where these thumbnails are being stored, in order to configure it in a way where you can effectively have a "ignore this folder" folder on Cloudinary itself. Similarly, we most likely want to review/finish/merge https://github.com/directus/directus/pull/8065 as part of this as well, so we can return links to the Cloudinary endpoint when querying the files from Directus' /files endpoint.

OK! That should be enough scope-creep for now. Lemme know what you think the MVP would be for your particular use case @don41382 🙂

**Reply by don41382** [reply]
Hi @rijkvanzanten,

Thanks for your fast reply! I started to do exactly what you suggested. But after reading your second post, I realized that Directus is acting as a proxy for all images. That is a no-go for my clients project and I guess, I can't wait on the PR #8065.

It's a shame! I really love how Directus is build. Keep up the good work.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
